Raven RawScreen Aspect Fix

andrikpowell opened the pull request:
So a thing that's bothered me for quite a while is that the "RawScreen"s that Raven uses have only been rendered using the "Not Adjusted" stretch setting... regardless of the stretch settings.

This PR fixes that and the issue mentioned here (#467). They are both related to the same issue in different ways. The problem with the specific issue, was that the "RawScreen" under "Fit to Width" was supposed to fit to width (imagine that), but code was set to prevent that.

Now the "RawScreen"s will look as they should for all three settings: "Not Adjusted", "Doom Format", and "Fit to Width". I will mention that the "RawScreen" code seems a bit WIP, so eventually there may be a better way to approach this further down the line.

One of the reasons the format was previously forced was probably due to the Raven Statusbar not being rendered properly under "Not Adjusted"... But the "RawScreen"s don't have that issue.

I haven't been able to fix the Raven Statusbar for "Not Adjusted", so I have it using the "Doom Format" stretch settings (just as DSDA Doom currently does).

Pedro-Beirao commented on prboom2/src/dsda/stretch.c:
```diff
@@ -293,6 +299,14 @@ void dsda_EvaluatePatchScale(void) {
+      wide_offset2x = SCREENWIDTH - WIDE_SCREENWIDTH;
+      wide_offset2y = SCREENHEIGHT - WIDE_SCREENHEIGHT;
+  }
+
```
Why did you make this change? It does the exact same thing
Althought instead of `render_stretch_hud++`, you could do `render_stretch_hud = patch_stretch_doom_format`

andrikpowell replied:
Original code changes every game asset that's "not adjusted" to become "doom format".

My code only does that to the statusbar as the statusbar doesn't display properly in "not adjusted".

But my code allows all other assets to be able to render as "not adjusted", which previously was not the case.

Pedro-Beirao replied:
Oh I see. The old code permanently changes the setting.
Missed that 😅

Pedro-Beirao replied:
But instead of doing `== 0`, do `== patch_stretch_not_adjusted`for readability

Pedro-Beirao commented on prboom2/src/dsda/stretch.c:
```diff
@@ -183,6 +183,15 @@ static void InitStretchParam(stretch_param_t* offsets, int stretch, enum patch_t
+    offsets->deltax1 = (SCREENWIDTH - WIDE_SCREENWIDTH) / 2;
+    offsets->deltax2 = (SCREENWIDTH - WIDE_SCREENWIDTH) / 2;
+  }
+
```
Do like the code below. Put an `if` before the `switch` statement that sets `stretch`to `patch_stretch_doom_format` if necessary

andrikpowell replied:
My thoughts process here is that this is a temporary fix, and by keeping this code isolated, it will be easier to cull once we figure out how to render the statusbar correctly in "not adjusted".

By putting an `if` in the `switch`, we are solidifying temp code that shouldn't be final.

Pedro-Beirao commented on prboom2/src/st_stuff.c:
```diff
@@ -416,6 +416,11 @@ void ST_SetScaledWidth(void)
+  // so let's use "patch_stretch_doom_format" settings
+  if (raven && render_stretch_hud == 0)
+      ST_SCALED_WIDTH  = width * WIDE_SCREENWIDTH / 320;
+
```
Same here

Pedro-Beirao commented on prboom2/src/v_video.c:
```diff
@@ -1440,11 +1440,29 @@ void V_DrawRawScreenSection(const char *lump_name, int source_offset, int dest_y
+      x_factor = 0;
+      y_factor = 0;
+      break;
+  }
```
No need for the `default` path

andrikpowell replied:
Only reason I added a default path was to fix GCC warnings.

Since I've started working on DSDA Doom code, it's been special in that GCC had no warnings in the continuous integration, and so whenever I work with the code, I try and keep it that way.

Pedro-Beirao replied:
All other instances of this switch statement are not using `default`. Thats why the comment.
And idk why the CI would not give you any warning, there are tons of warnings on my machine

andrikpowell replied:
What I'm particularly referencing is the GCC Linux and MSYS2 build log from the GitHub continuous integration.

When I started working on DSDA code these showed absolutely no warnings during the build process and so I tend to try and keep it that way.

I agree I wasn't crazy about using `default` there, I was just finding a way to squash a possible undefined warning relating to `x_factor` and `y_factor`.

andrikpowell commented:
Is this better?

fabiangreffrath commented:
Most of these changes appear redundant with code that is alrady there. It would be easier to implement a function (or macro) that returns `(raven && stretch == patch_stretch_not_adjusted) ? patch_stretch_doom_format : stretch` and use this in the `switch` statements instead of redefining in a later paragraph what was already checked before. This would also hold up the "temporary" aspect of these changes, as that function could later be changed to just return `stretch` (or get removed entirely).

andrikpowell commented:
> Most of these changes appear redundant with code that is alrady there. It would be easier to implement a function (or macro) that returns `(raven && stretch == patch_stretch_not_adjusted) ? patch_stretch_doom_format : stretch` and use this in the `switch` statements instead of redefining in a later paragraph what was already checked before. This would also hold up the "temporary" aspect of these changes, as that function could later be changed to just return `stretch` (or get removed entirely).

I think this is close to what you were suggesting? (check force push)

fabiangreffrath commented:
> I think this is close to what you were suggesting? (check force push)

Yes, except that it now uses the global `render_stretch_hud` in `InitStretchParam()` instead of the passed `stetch` parameter.

andrikpowell commented:
> Yes, except that it now uses the global `render_stretch_hud` in `InitStretchParam()` instead of the passed `stetch` parameter.

How do you wanna get around that?

Because the only way to force that `switch` statement to use `patch_stretch_doom_format` is to not use the passed `stretch` parameter.

I guess we could have a `if (stretch_hud != stretch)` then `stretch == stretch_hud` above?

fabiangreffrath commented:
> How do you wanna get around that?

```c
#define stretch_hud(a) ((raven && (a) == patch_stretch_not_adjusted) ? patch_stretch_doom_format : (a))
```

andrikpowell committed:
Render `V_DrawRawScreen` with correct stretch aspect
- Note that stbar in "Not Adjusted" aspect currently uses settings for "Doom Format"

andrikpowell commented:
> > How do you wanna get around that?
> 
> ```c
> #define stretch_hud(a) ((raven && (a) == patch_stretch_not_adjusted) ? patch_stretch_doom_format : (a))
> ```

oh fascinating! This is actually the first time I've seen a `#define` statement with arguments! That's actually really cool.

Force pushed new version with that included.

fabiangreffrath reviewed (approved):
Looks good now, thank you!

Pedro-Beirao commented:
Thank you both!
